File: packages/kotlin/src/test/kotlin/ai/crumbtrail/sdk/BehaviorTest.kt
package ai.crumbtrail.sdk

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

class SessionResolverTest {
    @Test
    fun `resumes a fresh session and refreshes its activity`() {
        val store = MemorySessionStore(PersistedSession("sess-old", 1_000))
        val resolved = CrumbtrailSessionResolver.resolve(store, idleMs = 5_000, now = 4_000) { "sess-new" }
        assertEquals("sess-old", resolved.id)
        assertEquals(4_000, resolved.lastActivity)
    }

    @Test
    fun `mints a fresh session once idle`() {
        // Resuming here would stitch today's bug onto last week's timeline.
        val store = MemorySessionStore(PersistedSession("sess-old", 1_000))
        val resolved = CrumbtrailSessionResolver.resolve(store, idleMs = 5_000, now = 99_000) { "sess-new" }
        assertEquals("sess-new", resolved.id)
    }

    @Test
    fun `resumes exactly at the idle boundary`() {
        val store = MemorySessionStore(PersistedSession("sess-old", 1_000))
        val resolved = CrumbtrailSessionResolver.resolve(store, idleMs = 5_000, now = 6_000) { "sess-new" }
        assertEquals("sess-old", resolved.id, "<= is the contract, not <")
    }

    @Test
    fun `persists the freshly minted id`() {
        val store = MemorySessionStore()
        CrumbtrailSessionResolver.resolve(store, idleMs = 5_000, now = 10) { "sess-new" }
        assertEquals("sess-new", store.read()?.id)
    }

    @Test
    fun `minted ids are distinct and correctly prefixed`() {
        val ids = (1..50).map { CrumbtrailSessionResolver.mintSessionId() }
        assertEquals(ids.size, ids.toSet().size)
        assertTrue(ids.all { it.startsWith("ses_") })
    }
}

/** Records every request and answers with a scripted status. */
private class StubHttpClient(
    private val status: Int = 200,
    private val throwsError: Boolean = false,
) : CrumbtrailHttpClient {
    data class Request(val url: String, val headers: Map<String, String>, val body: String)

    val requests = mutableListOf<Request>()

    override fun post(url: String, headers: Map<String, String>, body: ByteArray): Int {
        requests.add(Request(url, headers, String(body)))
        if (throwsError) throw java.io.IOException("offline")
        return status
    }
}

class TransportTest {
    private fun event() = CrumbtrailEvent(
        timestamp = 1_754_000_000_000,
        kind = CrumbtrailEventKind.ERROR,
        data = JsonValue.of("msg" to JsonValue.Str("boom")),
        sdk = CrumbtrailSdk.descriptor,
    )

    @Test
    fun `posts events to the contract path with the auth header`() {
        val client = StubHttpClient()
        CrumbtrailHttpTransport("https://api.crumbtrail.ai", "ctkey_abc", client)
            .sendEvents("sess-1", listOf(event()))

        val request = client.requests.single()
        assertEquals("https://api.crumbtrail.ai/api/events", request.url)
        assertEquals("application/json", request.headers["Content-Type"])
        assertEquals("ctkey_abc", request.headers["X-Crumbtrail-Auth"])
        assertTrue(request.body.contains("\"sessionId\":\"sess-1\""))
    }

    @Test
    fun `strips trailing slashes from the endpoint`() {
        val client = StubHttpClient()
        CrumbtrailHttpTransport("https://api.crumbtrail.ai///", client = client)
            .sendEvents("s", listOf(event()))
        // `//api/events` is a distinct, unrouted path on some gateways.
        assertEquals("https://api.crumbtrail.ai/api/events", client.requests.single().url)
    }

    @Test
    fun `omits the auth header entirely when no key is set`() {
        for (token in listOf(null, "")) {
            val client = StubHttpClient()
            CrumbtrailHttpTransport("https://api.crumbtrail.ai", token, client)
                .sendEvents("s", listOf(event()))
            // An empty value reads as a malformed credential, not as none.
            assertFalse(client.requests.single().headers.containsKey("X-Crumbtrail-Auth"))
        }
    }

    @Test
    fun `a refusal throws rather than counting as delivery`() {
        for (status in listOf(400, 401, 413, 429, 500)) {
            val client = StubHttpClient(status = status)
            val transport = CrumbtrailHttpTransport("https://api.crumbtrail.ai", client = client)
            val error = assertFailsWith<CrumbtrailDeliveryException.Refused> {
                transport.sendEvents("s", listOf(event()))
            }
            assertEquals(status, error.status)
            assertEquals(1, error.eventCount)
        }
    }

    @Test
    fun `a network failure is reported as unreachable`() {
        val client = StubHttpClient(throwsError = true)
        val transport = CrumbtrailHttpTransport("https://api.crumbtrail.ai", client = client)
        assertFailsWith<CrumbtrailDeliveryException.Unreachable> {
            transport.sendEvents("s", listOf(event()))
        }
    }

    @Test
    fun `an empty batch sends nothing`() {
        val client = StubHttpClient()
        CrumbtrailHttpTransport("https://api.crumbtrail.ai", client = client)
            .sendEvents("s", emptyList())
        assertTrue(client.requests.isEmpty())
    }

    @Test
    fun `session start and end use the contract paths`() {
        val client = StubHttpClient()
        val transport = CrumbtrailHttpTransport("https://api.crumbtrail.ai", client = client)
        transport.startSession("sess-1", JsonValue.of("service" to JsonValue.Str("app")))
        transport.endSession("sess-1")
        assertEquals(
            listOf(
                "https://api.crumbtrail.ai/api/session/start",
                "https://api.crumbtrail.ai/api/session/end",
            ),
            client.requests.map { it.url },
        )
    }
}

class EventQueueTest {
    private fun event(ms: Long) = CrumbtrailEvent(
        timestamp = ms,
        kind = CrumbtrailEventKind.ERROR,
        data = JsonValue.of(),
        sdk = CrumbtrailSdk.descriptor,
    )

    @Test
    fun `drops oldest once full and counts the drops`() {
        val queue = CrumbtrailEventQueue(capacity = 3)
        (1..5).forEach { queue.append(event(it.toLong())) }
        // Growing without bound would let a hot logging loop get the app killed
        // for memory — the SDK becoming the crash it was installed to explain.
        assertEquals(3, queue.size)
        assertEquals(2, queue.dropped)
        assertEquals(listOf(3L, 4L, 5L), queue.drain().map { it.timestamp })
    }

    @Test
    fun `drain empties the queue`() {
        val queue = CrumbtrailEventQueue(capacity = 10)
        queue.append(event(1))
        assertEquals(1, queue.drain().size)
        assertTrue(queue.drain().isEmpty())
    }

    @Test
    fun `requeue preserves chronological order`() {
        val queue = CrumbtrailEventQueue(capacity = 10)
        queue.append(event(3))
        // A retried batch appended at the back would sit after events that
        // happened later, inventing causality that never occurred.
        queue.requeue(listOf(event(1), event(2)))
        assertEquals(listOf(1L, 2L, 3L), queue.drain().map { it.timestamp })
    }

    @Test
    fun `requeue respects capacity`() {
        val queue = CrumbtrailEventQueue(capacity = 2)
        queue.append(event(9))
        queue.requeue(listOf(event(1), event(2), event(3)))
        assertEquals(listOf(3L, 9L), queue.drain().map { it.timestamp })
    }

    @Test
    fun `concurrent appends do not lose events`() {
        val queue = CrumbtrailEventQueue(capacity = 10_000)
        val threads = (1..8).map { worker ->
            Thread { (1..125).forEach { queue.append(event((worker * 1000 + it).toLong())) } }
        }
        threads.forEach { it.start() }
        threads.forEach { it.join() }
        assertEquals(1_000, queue.size)
    }
}

class RedactionTest {
    @Test
    fun `drops credential header values but keeps the names`() {
        val redacted = CrumbtrailRedaction.redactHeaders(
            mapOf(
                "Authorization" to "Bearer secret",
                "Cookie" to "session=abc",
                "X-API-Key" to "abc123",
                "Content-Type" to "application/json",
            )
        )
        // The name is diagnostic and harmless; only the value is dangerous.
        assertEquals("[REDACTED]", redacted["Authorization"])
        assertEquals("[REDACTED]", redacted["Cookie"])
        assertEquals("[REDACTED]", redacted["X-API-Key"])
        assertEquals("application/json", redacted["Content-Type"])
    }

    @Test
    fun `header matching ignores spelling and punctuation`() {
        for (spelling in listOf("X-API_Key", "x-api-key", "xApiKey", "X_API_KEY")) {
            assertTrue(
                CrumbtrailRedaction.isDeniedHeader(spelling),
                "$spelling must not slip through on spelling alone",
            )
        }
    }

    @Test
    fun `strips userinfo from urls`() {
        val redacted = CrumbtrailRedaction.redactUrl("https://user:pass@api.example.com/v1")
        assertFalse(redacted.contains("pass"))
        assertTrue(redacted.contains("api.example.com"))
    }

    @Test
    fun `redacts credential shaped query values but keeps the rest`() {
        val redacted = CrumbtrailRedaction.redactUrl(
            "https://api.example.com/v1?page=2&access_token=abc123&q=shoes"
        )
        assertFalse(redacted.contains("abc123"))
        // The rest of the query is what makes the request diagnosable.
        assertTrue(redacted.contains("page=2"))
        assertTrue(redacted.contains("q=shoes"))
    }

    @Test
    fun `drops the fragment where oauth tokens land`() {
        val redacted = CrumbtrailRedaction.redactUrl(
            "https://app.example.com/callback#access_token=abc123"
        )
        assertFalse(redacted.contains("abc123"))
    }

    @Test
    fun `strings without a scheme and host keep only their path`() {
        // URI is lenient, so "did it parse" proves nothing. The query-key
        // heuristic is only trustworthy on a real URL, so everything else has
        // its query dropped wholesale rather than scanned.
        assertEquals("/v1/orders", CrumbtrailRedaction.redactUrl("/v1/orders?access_token=abc"))
        assertEquals("[REDACTED]", CrumbtrailRedaction.redactUrl(""))
        assertFalse(CrumbtrailRedaction.redactUrl("not a url?token=abc123").contains("abc123"))
    }
}

class CrumbtrailTest {
    private class CapturingTransport(private val failWith: Int? = null) : CrumbtrailTransport {
        val batches = mutableListOf<List<CrumbtrailEvent>>()
        var startedSession: String? = null
        var endedSession: String? = null

        override fun startSession(id: String, metadata: JsonValue) { startedSession = id }
        override fun endSession(id: String) { endedSession = id }
        override fun sendEvents(sessionId: String, events: List<CrumbtrailEvent>) {
            batches.add(events)
            failWith?.let { throw CrumbtrailDeliveryException.Refused(it, events.size) }
        }
    }

    private fun config(batchSize: Int = 1000) = CrumbtrailConfig(
        endpoint = "https://api.crumbtrail.ai",
        service = "app",
        flushBatchSize = batchSize,
        // No timer: a background flush would race every assertion below. The
        // tests below also pass `CrumbtrailInlineDelivery`, which has no timer
        // of its own and makes each delivery observable the moment it is asked
        // for. Production never uses it — see CrumbtrailDelivery.
        flushIntervalSeconds = 0,
    )

    @Test
    fun `announces the session and emits a startup environment snapshot`() {
        val transport = CapturingTransport()
        val logger = Crumbtrail(config(), transport, MemorySessionStore(), delivery = CrumbtrailInlineDelivery)

        assertEquals(logger.sessionId, transport.startedSession)
        logger.flush()
        assertEquals("env", transport.batches.single().single().kind)
    }

    @Test
    fun `a refusal becomes a declared gap instead of a silent loss`() {
        val transport = CapturingTransport(failWith = 413)
        val logger = Crumbtrail(config(), transport, MemorySessionStore(), delivery = CrumbtrailInlineDelivery)

        logger.flush()

        // A session missing events must not read like a session where nothing
        // happened, so the hole is a fact in the timeline rather than an absence.
        assertEquals(1, logger.gaps.size)
        assertEquals("refused-413", logger.gaps.single().reason)
    }

    @Test
    fun `an unreachable endpoint requeues rather than dropping`() {
        val transport = object : CrumbtrailTransport {
            var attempts = 0
            override fun startSession(id: String, metadata: JsonValue) = Unit
            override fun endSession(id: String) = Unit
            override fun sendEvents(sessionId: String, events: List<CrumbtrailEvent>) {
                attempts++
                if (attempts == 1) throw CrumbtrailDeliveryException.Unreachable(events.size)
            }
        }
        val logger = Crumbtrail(config(), transport, MemorySessionStore(), delivery = CrumbtrailInlineDelivery)

        logger.flush()
        assertTrue(logger.gaps.isEmpty(), "a network failure is retried, not declared lost")
        logger.flush()
        assertEquals(2, transport.attempts)
    }

    @Test
    fun `flushes automatically once the batch size is reached`() {
        val transport = CapturingTransport()
        val logger = Crumbtrail(
            config(batchSize = 2), transport, MemorySessionStore(), delivery = CrumbtrailInlineDelivery
        )
        // The startup env event is already buffered, so one more reaches 2.
        logger.addEvent(CrumbtrailEventKind.ERROR, JsonValue.of("msg" to JsonValue.Str("x")))
        assertEquals(1, transport.batches.size)
    }

    @Test
    fun `recordRequest redacts the url before it is buffered`() {
        val transport = CapturingTransport()
        val logger = Crumbtrail(config(), transport, MemorySessionStore(), delivery = CrumbtrailInlineDelivery)

        logger.recordRequest(
            url = "https://api.example.com/v1?access_token=abc123",
            method = "get",
            status = 402,
            durationMs = 318,
        )
        logger.flush()

        val json = transport.batches.single().first { it.kind == "net" }.toJson().toJson()
        assertFalse(json.contains("abc123"))
        assertTrue(json.contains("\"method\":\"GET\""))
        assertTrue(json.contains("\"ok\":false"))
    }

    @Test
    fun `stop flushes and closes the session`() {
        val transport = CapturingTransport()
        val logger = Crumbtrail(config(), transport, MemorySessionStore(), delivery = CrumbtrailInlineDelivery)

        logger.stop()
        assertEquals(logger.sessionId, transport.endedSession)
        // Events added after stop are dropped rather than queued forever.
        logger.addEvent(CrumbtrailEventKind.ERROR, JsonValue.of())
        assertEquals(1, transport.batches.size)
    }

    @Test
    fun `activity is touched so a resumed session does not expire mid use`() {
        val store = MemorySessionStore()
        var clock = 1_000L
        val logger = Crumbtrail(
            config(), CapturingTransport(), store, delivery = CrumbtrailInlineDelivery, clock = { clock }
        )
        clock = 50_000
        logger.addEvent(CrumbtrailEventKind.ERROR, JsonValue.of())
        assertEquals(50_000, store.read()?.lastActivity)
    }
}

File: packages/flutter/test/behavior_test.dart
import 'package:crumbtrail_flutter/crumbtrail_flutter.dart';
import 'package:flutter_test/flutter_test.dart';

/// Records what it was asked to send, and fails on command.
class FakeTransport implements CrumbtrailTransport {
  FakeTransport({this.failure});

  /// Thrown by the next `sendEvents`, if set. Cleared after it throws once, so
  /// a test can assert on the retry that follows.
  CrumbtrailDeliveryException Function(int count)? failure;

  final List<List<CrumbtrailEvent>> batches = [];
  final List<String> started = [];
  final List<String> ended = [];
  Map<String, Object?>? startMetadata;

  List<CrumbtrailEvent> get allEvents =>
      batches.expand((batch) => batch).toList();

  @override
  Future<void> startSession(String id, Map<String, Object?> metadata) async {
    started.add(id);
    startMetadata = metadata;
  }

  @override
  Future<void> sendEvents(String sessionId, List<CrumbtrailEvent> events) async {
    final failNow = failure;
    if (failNow != null) {
      failure = null;
      throw failNow(events.length);
    }
    batches.add(events);
  }

  @override
  Future<void> endSession(String id) async => ended.add(id);
}

const _endpoint = 'https://ingest.example.com';

Crumbtrail buildLogger({
  CrumbtrailTransport? transport,
  CrumbtrailSessionStore? store,
  CrumbtrailConfig? config,
  int? now,
}) {
  return Crumbtrail(
    config: config ??
        const CrumbtrailConfig(
          endpoint: _endpoint,
          // Large enough that no test flushes by accident on batch size.
          flushBatchSize: 1000,
        ),
    transport: transport ?? FakeTransport(),
    store: store,
    deviceInfo: const {'os': 'test'},
    // A periodic timer in a unit test is a source of flake, not coverage.
    startTimer: false,
    clock: now == null ? null : () => now,
  );
}

void main() {
  group('session resolution', () {
    test('a recent session is resumed across launches', () {
      final store = MemorySessionStore(
        const PersistedSession(id: 'ses_previous', lastActivity: 1000000),
      );
      final logger = buildLogger(store: store, now: 1000000 + 60000);
      // Within the 30 minute idle window.
      expect(logger.sessionId, 'ses_previous');
    });

    test('a stale session is not resumed', () {
      final store = MemorySessionStore(
        const PersistedSession(id: 'ses_previous', lastActivity: 1000000),
      );
      // One millisecond past the idle window. Resuming here would stitch
      // yesterday's timeline onto today's bug.
      final logger = buildLogger(
        store: store,
        now: 1000000 + (30 * 60 * 1000) + 1,
      );
      expect(logger.sessionId, isNot('ses_previous'));
      expect(logger.sessionId, startsWith('ses_'));
    });

    test('the boundary itself resumes', () {
      final store = MemorySessionStore(
        const PersistedSession(id: 'ses_previous', lastActivity: 1000000),
      );
      final logger = buildLogger(
        store: store,
        now: 1000000 + (30 * 60 * 1000),
      );
      expect(logger.sessionId, 'ses_previous');
    });

    test('a corrupt stored session starts a fresh one rather than throwing', () {
      // Startup is the worst possible place to throw: the app would crash
      // because a telemetry SDK could not read its own scratch file.
      expect(PersistedSession.tryParse('{not json'), isNull);
      expect(PersistedSession.tryParse('{"lastActivity":1}'), isNull);
      expect(PersistedSession.tryParse('{"id":""}'), isNull);
      expect(PersistedSession.tryParse(null), isNull);
      expect(PersistedSession.tryParse('[]'), isNull);
    });

    test('a session missing lastActivity is treated as ancient', () {
      final parsed = PersistedSession.tryParse('{"id":"ses_x"}');
      expect(parsed, isNotNull);
      expect(parsed!.lastActivity, 0);
    });

    test('activity keeps a session from expiring mid-use', () {
      final store = MemorySessionStore();
      final logger = buildLogger(store: store, now: 5000000);
      logger.addEvent(CrumbtrailEventKind.console, {'lv': 'log'});
      expect(store.read()!.lastActivity, 5000000);
      expect(store.read()!.id, logger.sessionId);
    });

    test('minted ids carry a UTC stamp', () {
      final id = CrumbtrailSessionResolver.mintSessionId(1754000000000);
      expect(id, startsWith('ses_20250731_'));
      expect(id.split('_').last.length, 12);
    });

    test('minted ids do not collide', () {
      final ids = List.generate(
        200,
        (_) => CrumbtrailSessionResolver.mintSessionId(1754000000000),
      ).toSet();
      expect(ids.length, 200);
    });
  });

  group('delivery', () {
    test('a refusal is recorded as a gap, not retried', () async {
      final transport = FakeTransport(
        failure: (count) => CrumbtrailRefused(413, count),
      );
      final logger = buildLogger(transport: transport);
      logger.addEvent(CrumbtrailEventKind.console, {'lv': 'log'});
      await logger.flush();

      expect(logger.gaps, hasLength(1));
      expect(logger.gaps.single.reason, 'refused-413');
      // The server already answered. Retrying the identical batch would be
      // refused identically, and a silent drop reads as "nothing happened".
      expect(transport.batches, isEmpty);

      await logger.flush();
      expect(transport.batches, isEmpty);
    });

    test('an unreachable endpoint is retried on the next flush', () async {
      final transport = FakeTransport(
        failure: (count) => CrumbtrailUnreachable(count),
      );
      final logger = buildLogger(transport: transport);
      logger.addEvent(CrumbtrailEventKind.console, {'lv': 'log'});

      await logger.flush();
      expect(transport.batches, isEmpty);
      expect(logger.gaps, isEmpty);

      await logger.flush();
      expect(transport.allEvents, hasLength(2)); // env snapshot + console
    });

    test('a retried batch keeps its place in the timeline', () async {
      final transport = FakeTransport(
        failure: (count) => CrumbtrailUnreachable(count),
      );
      final logger = buildLogger(transport: transport);
      logger.addEvent(CrumbtrailEventKind.console, {'seq': 1});

      await logger.flush(); // fails, requeues at the front
      logger.addEvent(CrumbtrailEventKind.console, {'seq': 2});
      await logger.flush();

      final sequence = transport.allEvents
          .where((event) => event.kind == 'con')
          .map((event) => event.data['seq'])
          .toList();
      // Re-appending at the back would put the retried batch after events that
      // happened later, inventing causality that never occurred.
      expect(sequence, [1, 2]);
    });

    test('flushing an empty queue does not call the transport', () async {
      final transport = FakeTransport();
      final logger = buildLogger(transport: transport);
      await logger.flush();
      transport.batches.clear();
      await logger.flush();
      expect(transport.batches, isEmpty);
    });

    test('stop flushes, ends the session, and stops recording', () async {
      final transport = FakeTransport();
      final logger = buildLogger(transport: transport);
      logger.addEvent(CrumbtrailEventKind.console, {'lv': 'log'});
      await logger.stop();

      expect(transport.ended, [logger.sessionId]);
      expect(transport.allEvents, hasLength(2));

      logger.addEvent(CrumbtrailEventKind.console, {'lv': 'after'});
      await logger.flush();
      expect(transport.allEvents, hasLength(2));
    });

    test('session start announces the service name', () async {
      final transport = FakeTransport();
      buildLogger(
        transport: transport,
        config: const CrumbtrailConfig(
          endpoint: _endpoint,
          service: 'checkout-app',
          flushBatchSize: 1000,
        ),
      );
      // One ingest key covers a whole project, so without this every app in it
      // arrives as an anonymous sender.
      await Future<void>.delayed(Duration.zero);
      expect(transport.startMetadata?['service'], 'checkout-app');
      expect(transport.startMetadata?['platform'], 'flutter');
    });

    test('reaching the batch size flushes without waiting for the timer',
        () async {
      final transport = FakeTransport();
      final logger = buildLogger(
        transport: transport,
        config: const CrumbtrailConfig(
          endpoint: _endpoint,
          flushBatchSize: 3,
        ),
      );
      logger.addEvent(CrumbtrailEventKind.console, {'n': 1});
      logger.addEvent(CrumbtrailEventKind.console, {'n': 2});
      await Future<void>.delayed(Duration.zero);
      expect(transport.allEvents, hasLength(3)); // env + 2
    });
  });

  group('queue bounds', () {
    test('the oldest events are dropped and counted', () {
      final queue = CrumbtrailEventQueue(capacity: 3);
      for (var index = 0; index < 5; index++) {
        queue.append(CrumbtrailEvent.of(
          timestamp: index,
          kind: CrumbtrailEventKind.console,
          data: {'n': index},
          sdk: crumbtrailSdk,
        ));
      }
      expect(queue.length, 3);
      expect(queue.dropped, 2);
      // The most recent window is the one the bug is in.
      expect(queue.drain().map((event) => event.data['n']), [2, 3, 4]);
    });

    test('a requeue larger than the queue keeps the newest of the batch', () {
      final queue = CrumbtrailEventQueue(capacity: 2);
      queue.requeue([
        for (var index = 0; index < 4; index++)
          CrumbtrailEvent.of(
            timestamp: index,
            kind: CrumbtrailEventKind.console,
            data: {'n': index},
            sdk: crumbtrailSdk,
          ),
      ]);
      expect(queue.drain().map((event) => event.data['n']), [2, 3]);
      expect(queue.dropped, 2);
    });

    test('capacity below one is clamped rather than dividing by zero', () {
      final queue = CrumbtrailEventQueue(capacity: 0);
      queue.append(CrumbtrailEvent.of(
        timestamp: 1,
        kind: CrumbtrailEventKind.console,
        data: const {},
        sdk: crumbtrailSdk,
      ));
      expect(queue.length, 1);
    });

    test('drain empties the queue', () {
      final queue = CrumbtrailEventQueue(capacity: 10);
      queue.append(CrumbtrailEvent.of(
        timestamp: 1,
        kind: CrumbtrailEventKind.console,
        data: const {},
        sdk: crumbtrailSdk,
      ));
      expect(queue.drain(), hasLength(1));
      expect(queue.drain(), isEmpty);
    });
  });

  group('redaction', () {
    test('credential-shaped header values go, names stay', () {
      final redacted = CrumbtrailRedaction.redactHeaders({
        'Authorization': 'Bearer abc.def.ghi',
        'X-API_Key': 'sk-live-1234',
        'x-csrf-token': 'nonce',
        'Cookie': 'sid=1',
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      });
      // "The request carried an Authorization header" is diagnostic and
      // harmless; only the value is dangerous.
      expect(redacted['Authorization'], '[REDACTED]');
      expect(redacted['X-API_Key'], '[REDACTED]');
      expect(redacted['x-csrf-token'], '[REDACTED]');
      expect(redacted['Cookie'], '[REDACTED]');
      expect(redacted['Content-Type'], 'application/json');
      expect(redacted['Accept'], 'application/json');
    });

    test('spelling variants cannot slip a credential through', () {
      for (final name in [
        'x-api-key',
        'X_API_KEY',
        'xApiKey',
        'X.Api.Key',
        'AUTHORIZATION',
      ]) {
        expect(
          CrumbtrailRedaction.isDeniedHeader(name),
          isTrue,
          reason: '$name should be denied',
        );
      }
      expect(CrumbtrailRedaction.isDeniedHeader('accept-language'), isFalse);
    });

    test('userinfo and fragment never leave the device', () {
      final url = CrumbtrailRedaction.redactUrl(
        'https://user:hunter2@api.example.com/v1/orders#access_token=abc',
      );
      expect(url, isNot(contains('hunter2')));
      expect(url, isNot(contains('user')));
      // The fragment is where apps park an access token after an OAuth redirect.
      expect(url, isNot(contains('access_token')));
      expect(url, 'https://api.example.com/v1/orders');
    });

    test('credential-shaped query values are dropped, ordinary ones kept', () {
      final url = CrumbtrailRedaction.redactUrl(
        'https://api.example.com/search?q=shoes&api_key=sk-live&page=2',
      );
      expect(url, contains('q=shoes'));
      expect(url, contains('page=2'));
      expect(url, isNot(contains('sk-live')));
      // Literally, not percent encoded. Ingest matches on the marker string, so
      // a `%5BREDACTED%5D` here reads as an ordinary value on the other side.
      expect(url, contains('[REDACTED]'));
      expect(url, 'https://api.example.com/search?q=shoes&api_key=[REDACTED]&page=2');
    });

    test('a non-URL keeps only its path, never its query', () {
      // Uri.parse is lenient, so "did it parse" is not a safety check: a
      // schemeless string could otherwise carry a token straight through.
      expect(
        CrumbtrailRedaction.redactUrl('not a url?token=secret'),
        isNot(contains('secret')),
      );
      expect(CrumbtrailRedaction.redactUrl(''), '[REDACTED]');
    });

    test('the port is preserved', () {
      expect(
        CrumbtrailRedaction.redactUrl('http://localhost:8080/api/health'),
        'http://localhost:8080/api/health',
      );
    });

    test('recorded requests are redacted before they are queued', () async {
      final transport = FakeTransport();
      final logger = buildLogger(transport: transport);
      logger.recordRequest(
        url: 'https://api.example.com/v1/orders?token=sk-live-1234',
        method: 'post',
        status: 402,
        durationMs: 318,
      );
      await logger.flush();

      final event =
          transport.allEvents.firstWhere((event) => event.kind == 'net');
      expect(event.data['url'], isNot(contains('sk-live-1234')));
      expect(event.data['method'], 'POST');
      expect(event.data['ok'], isFalse);
      expect(event.data['dur'], 318);
    });

    test('a 2xx status records ok', () async {
      final transport = FakeTransport();
      final logger = buildLogger(transport: transport);
      logger.recordRequest(
        url: 'https://api.example.com/v1/orders',
        method: 'GET',
        status: 204,
        durationMs: 12,
      );
      await logger.flush();
      final event =
          transport.allEvents.firstWhere((event) => event.kind == 'net');
      expect(event.data['ok'], isTrue);
    });
  });

  group('recording', () {
    test('a caught error is not marked fatal', () async {
      final transport = FakeTransport();
      final logger = buildLogger(transport: transport);
      logger.recordError(StateError('boom'), StackTrace.current);
      await logger.flush();
      final event =
          transport.allEvents.firstWhere((event) => event.kind == 'err');
      // The app survived. Marking it fatal would inflate every crash metric.
      expect(event.data['fatal'], isFalse);
      expect(event.data['msg'], contains('boom'));
    });

    test('an environment snapshot is emitted at startup', () async {
      final transport = FakeTransport();
      final logger = buildLogger(transport: transport);
      await logger.flush();
      final event =
          transport.allEvents.firstWhere((event) => event.kind == 'env');
      expect(event.data['kind'], 'snapshot');
    });

    test('the environment collector can be turned off', () async {
      final transport = FakeTransport();
      final logger = buildLogger(
        transport: transport,
        config: const CrumbtrailConfig(
          endpoint: _endpoint,
          flushBatchSize: 1000,
          collectors: CrumbtrailCollectors(environment: false),
        ),
      );
      await logger.flush();
      expect(transport.allEvents, isEmpty);
    });

    test('describePlatform collects nothing that identifies a person', () {
      final described = Crumbtrail.describePlatform();
      expect(described.keys, isNot(contains('deviceId')));
      expect(described.keys, isNot(contains('advertisingId')));
      expect(described['os'], isNotNull);
    });
  });

  group('transport wiring', () {
    test('a trailing slash does not produce a double-slashed path', () async {
      final calls = <String>[];
      final transport = CrumbtrailHttpTransport(
        endpoint: 'https://ingest.example.com/',
        ingestKey: 'ctkey_test',
        client: _RecordingClient(calls),
      );
      await transport.sendEvents('ses_1', [
        CrumbtrailEvent.of(
          timestamp: 1,
          kind: CrumbtrailEventKind.console,
          data: const {},
          sdk: crumbtrailSdk,
        ),
      ]);
      // Some gateways route `//api/events` as a distinct, unrouted path.
      expect(calls.single, 'https://ingest.example.com/api/events');
    });

    test('a non-2xx response is a refusal, not a delivery', () async {
      final transport = CrumbtrailHttpTransport(
        endpoint: _endpoint,
        client: _StatusClient(429),
      );
      await expectLater(
        transport.sendEvents('ses_1', [
          CrumbtrailEvent.of(
            timestamp: 1,
            kind: CrumbtrailEventKind.console,
            data: const {},
            sdk: crumbtrailSdk,
          ),
        ]),
        throwsA(isA<CrumbtrailRefused>()),
      );
    });

    test('a thrown client is unreachable, which is retryable', () async {
      final transport = CrumbtrailHttpTransport(
        endpoint: _endpoint,
        client: _ThrowingClient(),
      );
      await expectLater(
        transport.sendEvents('ses_1', [
          CrumbtrailEvent.of(
            timestamp: 1,
            kind: CrumbtrailEventKind.console,
            data: const {},
            sdk: crumbtrailSdk,
          ),
        ]),
        throwsA(isA<CrumbtrailUnreachable>()),
      );
    });

    test('session start and end failures never surface to the host app', () async {
      final transport = CrumbtrailHttpTransport(
        endpoint: _endpoint,
        client: _ThrowingClient(),
      );
      // Capture that cannot announce itself still captures; ingest creates the
      // session lazily from the first batch.
      await transport.startSession('ses_1', const {});
      await transport.endSession('ses_1');
    });

    test('the ingest key is sent as a header, an empty one is omitted',
        () async {
      final withKey = <Map<String, String>>[];
      await CrumbtrailHttpTransport(
        endpoint: _endpoint,
        ingestKey: 'ctkey_abc',
        client: _HeaderClient(withKey),
      ).endSession('ses_1');
      expect(withKey.single['X-Crumbtrail-Auth'], 'ctkey_abc');

      final withoutKey = <Map<String, String>>[];
      await CrumbtrailHttpTransport(
        endpoint: _endpoint,
        ingestKey: '',
        client: _HeaderClient(withoutKey),
      ).endSession('ses_1');
      // An empty value reads to the server as a malformed credential rather
      // than as none at all.
      expect(withoutKey.single.containsKey('X-Crumbtrail-Auth'), isFalse);
    });

    test('an empty batch is not posted', () async {
      final calls = <String>[];
      await CrumbtrailHttpTransport(
        endpoint: _endpoint,
        client: _RecordingClient(calls),
      ).sendEvents('ses_1', const []);
      expect(calls, isEmpty);
    });
  });
}

class _RecordingClient implements CrumbtrailHttpClient {
  _RecordingClient(this.calls);
  final List<String> calls;

  @override
  Future<int> post(String url, Map<String, String> headers, String body) async {
    calls.add(url);
    return 200;
  }
}

class _HeaderClient implements CrumbtrailHttpClient {
  _HeaderClient(this.seen);
  final List<Map<String, String>> seen;

  @override
  Future<int> post(String url, Map<String, String> headers, String body) async {
    seen.add(headers);
    return 200;
  }
}

class _StatusClient implements CrumbtrailHttpClient {
  _StatusClient(this.status);
  final int status;

  @override
  Future<int> post(String url, Map<String, String> headers, String body) async =>
      status;
}

class _ThrowingClient implements CrumbtrailHttpClient {
  @override
  Future<int> post(String url, Map<String, String> headers, String body) async {
    throw const SocketExceptionStub();
  }
}

class SocketExceptionStub implements Exception {
  const SocketExceptionStub();
}
